### DISCUSSION TITLE
Proper way to use ModelTracker?

### DISCUSSION THREAD

**Discussion by npwoods** [slint-ui/slint#5460] [Q&A]
What is the proper way for user code to observe a `Model`, presumably using the `dyn ModelTracker` returned by `Model::model_tracker()`?

I was trying to chain multiple models together (a common pattern in Qt), and I also intend to write unit tests for my implementations of `Model`, and presumably I would need to observe models to properly exercise them in unit tests.

Thanks!

**Comment by ogoffart** [comment]
There is currently no public API to observe a model. 
There is private API for that which is i_slint_core::model::ModelChangeListenerContainer
What was your precise usecase? There is already MapModel and FilterModel that can be used for chaining.

**Comment by npwoods** [comment]
In my case I was trying to split out a non trivial model into two models, specifically an abstract (and crude) tree mechanism similar to what is done in the `cargo-ui` example, and another model with the actual backing data (neither `MapModel` nor `FilterModel` really did what I wanted).  I concluded that I was trying to "_cut against the grain_" and opted to do a form of composition that did not involve distinct implementations of `Model` - which is fine.

I will say that ideally I could write unit tests against my implementations of `Model` (e.g. - exercise the model, and assert that `row_added` is called in the fashion I expected).  If there is a kosher way to do this, that would be awesome!
